libs/mpi/tests/unit/mpi_ring_async_executor.cpp

```diff
@@ -74,63 +89,84 @@ int hpx_main(hpx::program_options::variables_map& vm)
+                    if (rank > 0)
+                    {
+                        // send the incremented token to the next rank
+                        ++tokens[i];
+                        hpx::future<int> f_send = hpx::async(
+                            exec, MPI_Isend, &tokens[i], 1, MPI_INT, rank_to, i);
+                        // when the send completes
+                        f_send.then([=, &tokens, &counter](auto&&) {
+                            msg_send(rank, size, rank_to, rank_from, tokens[i], i);
+                            // ranks > 0 are done when they have sent their token
+                            --counter;
+                        });
+                    }
+                    else
+                    {
+                        // rank 0 is done when it receives its token
+                        --counter;
+                    }
+                });
+
+                // rank 0 starts the process with a send
+                if (rank == 0)
                 {
-                    // send the incremented token to the next rank
-                    ++tokens[i];
-                    hpx::future<void> f_send = hpx::async(
+                    auto f_send = hpx::async(
                         exec, MPI_Isend, &tokens[i], 1, MPI_INT, rank_to, i);
-                    // when the send completes
                     f_send.then([=, &tokens, &counter](auto&&) {
                         msg_send(rank, size, rank_to, rank_from, tokens[i], i);
-                        // ranks > 0 are done when they have sent their token
-                        --counter;
                     });
                 }
-                else
-                {
-                    // rank 0 is done when it receives its token
-                    --counter;
-                }
-            });
+                --remainder;
+            }
 
-            // rank 0 starts the process with a send
-            if (rank == 0)
+            // Our simple counter should reach zero when all send/recv pairs are done
+            hpx::mpi::experimental::wait([&]() { return counter != 0; });
+            if (rank == 0 && output)
             {
-                auto f_send = hpx::async(
-                    exec, MPI_Isend, &tokens[i], 1, MPI_INT, rank_to, i);
-                f_send.then([=, &tokens, &counter](auto&&) {
-                    msg_send(rank, size, rank_to, rank_from, tokens[i], i);
-                });
+                std::cout << "remainder " << remainder << " j " << j << std::endl;
             }
         }
 
-        // Our simple counter should reach zero when all send/recv pairs are done
-        hpx::mpi::experimental::wait([&]() { return counter != 0; });
-
-        // let the user polling go out of scope
-    }
+        // This is needed to make sure that one rank does not shut down
+        // before others have completed. MPI does not handle that well.
+        MPI_Barrier(MPI_COMM_WORLD);
```

Comment:
This could probably go away if we handled the futures returned from the `.then()` above explicitly.

Reply:
This is really there for 2 reasons :
Users will look at our unit tests as a base to develop their own codes - adding a barrier before shutting down the runtime is good practice and I put it here for that reason - but also because, if rank 0 has received all its messages, sent all its sends and is "finished" all its futures will be ready and it can complete and exit - but if it tries to exit whilst other ranks are still doing exchanges, then 99% of the time there will be a problem. Adding this barrier is there to guarantee this does not happen and show users the best practice.

Reply:
Would having an MPI barrier in the HPX code right before MPI_finalize is called be sufficient? I would not consider using MPI_Barrier in user code as 'good practice'. It makes me cringe when I see it.

Reply:
Adding a barrier before finalize would certainly help, but only if the user allowed hpx to do the init. If they handle init/finalize themselves then it won't help. I do agree that barriers are bad, but a barrier at shutdown is not as evil as one after every iteration - and lets nioot forget that inside hps there are things like `hpx::big_boot_barrier` that serve exactly the same purpose as this except at startup instead of shutdown. I'd prefer to leave it in.

Reply:
If the users do MPI_init/MPI_finalize themselves we can expect them to handle the barrier in the end anyways as they will do other things requiring to do that anyways. I'd vote for adding an MPI_Barrier before we call MPI_finalize and remove it here.

Reply:
> and lets not forget that inside hpx there are things like hpx::big_boot_barrier that serve exactly the same purpose

Yes. However those are inside HPX and not in application code. The same would be true for adding an MPI_barrier before calling MPI_finalize. I have no problem if the user decides to add a barrier in their code wherever they want. Mostly I would like to avoid us requiring for the user to do that on our behalf.

Reply:
This is an application example and not part of hpx internals, the example calls mpi_init explicitly on start (as most applications will and does not use any hpx mpi initialization). The MPI_Barrier is not part of any hpx requirement. This test does work without the barrier (does not crash), because I've made a point of adding a wait until all requests are complete - but if the user does not wait on all requests or futures as we do with our counter, then it will probably crash. I am happy to comment out the barrier, but with notes to say that not waiting for all futures to complete is very bad (especially if they mix normal mpi calls with futurized ones). This is supposed to be an example to show users how to safely use the mpi integration and if we don't mention it, they will have crashes and we'll just get people filing issues and spamming IRC. I will also add a note to the docs about it.

Reply:
Ok, then let's move the barrier into `main()` (after `hpx::init` returned) to disentangle the HPX part from the MPI code.

However, I still think that dropping the futures on the floor is a bad thing, but this might be for a separate PR.